torch_sim/constraints.py

```diff
@@ -0,0 +1,406 @@
+"""Constraints for molecular dynamics simulations.
+
+This module implements constraints inspired by ASE's constraint system,
+adapted for the torch-sim framework with support for batched operations
+and PyTorch tensors.
+
+The constraints affect degrees of freedom counting and modify forces, momenta,
+and positions during MD simulations.
+"""
+
+from __future__ import annotations
+
+import warnings
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Any
+
+import torch
+
+
+if TYPE_CHECKING:
+    from torch_sim.state import SimState
+
+
+class FixConstraint(ABC):
+    """Base class for all constraints in torch-sim.
+
+    This is the abstract base class that all constraints must inherit from.
+    It defines the interface that constraints must implement to work with
+    the torch-sim MD system.
+    """
+
+    @abstractmethod
+    def get_removed_dof(self, state: SimState) -> torch.Tensor:
+        """Get the number of degrees of freedom removed by this constraint.
+
+        Args:
+            state: The simulation state
+
+        Returns:
+            Number of degrees of freedom removed by this constraint
+        """
+
+    @abstractmethod
+    def adjust_positions(self, state: SimState, new_positions: torch.Tensor) -> None:
+        """Adjust positions to satisfy the constraint.
+
+        This method should modify new_positions in-place to ensure the
+        constraint is satisfied.
+
+        Args:
+            state: Current simulation state
+            new_positions: Proposed new positions to be adjusted
+        """
+
+    def adjust_momenta(self, state: SimState, momenta: torch.Tensor) -> None:
+        """Adjust momenta to satisfy the constraint.
+
+        This method should modify momenta in-place to ensure the constraint
+        is satisfied. By default, it calls adjust_forces with the momenta.
+
+        Args:
+            state: Current simulation state
+            momenta: Momenta to be adjusted
+        """
+        # Default implementation: treat momenta like forces
+        self.adjust_forces(state, momenta)
+
+    @abstractmethod
+    def adjust_forces(self, state: SimState, forces: torch.Tensor) -> None:
+        """Adjust forces to satisfy the constraint.
+
+        This method should modify forces in-place to ensure the constraint
+        is satisfied.
+
+        Args:
+            state: Current simulation state
+            forces: Forces to be adjusted
+        """
+
+    def copy(self) -> FixConstraint:
+        """Create a copy of this constraint.
+
+        Returns:
+            A new instance of this constraint with the same parameters
+        """
+        return type(self)(**self.__dict__)
+
+    def todict(self) -> dict[str, Any]:
+        """Convert constraint to dictionary representation.
+
+        Returns:
+            Dictionary representation of the constraint
+        """
+        return {"name": self.__class__.__name__, "kwargs": self.__dict__.copy()}
```

Comment:
I'd get rid of this too. `asdict` is a built-in for `dataclasses` and users can add a name themselves if they want.